File: curriculum/units/M04/code/typescript/traceHarness.ts
export type HarnessMessage =
  | { kind: 'user'; text: string }
  | { kind: 'assistant'; text: string }

export type TraceEventInput =
  | { type: 'call.entered'; source: string; target: string }
  | { type: 'state.mutated'; owner: string; field: string; size: number }
  | { type: 'view.snapshotted'; owner: string; field: string; size: number }
  | { type: 'event.yielded'; source: string; eventKind: HarnessMessage['kind'] }
  | { type: 'branch.skipped'; branch: string; reason: string }
  | { type: 'call.failed'; source: string; message: string }

export type TraceEvent = TraceEventInput & { seq: number }

export class TraceLog {
  readonly events: TraceEvent[] = []
  #nextSeq = 1

  record(event: TraceEventInput): void {
    this.events.push({ ...event, seq: this.#nextSeq++ } as TraceEvent)
  }

  observedCall(source: string, target: string): boolean {
    return this.events.some(
      event =>
        event.type === 'call.entered' &&
        event.source === source &&
        event.target === target,
    )
  }
}

export type ProcessInputResult = {
  messages: HarnessMessage[]
  shouldQuery: boolean
}

export type ProcessInput = (
  prompt: string,
  history: readonly HarnessMessage[],
) => Promise<ProcessInputResult>

export type QueryStream = (
  messages: readonly HarnessMessage[],
) => AsyncIterable<HarnessMessage>

export class TraceableEngine {
  #messages: HarnessMessage[]
  readonly #processInput: ProcessInput
  readonly #queryStream: QueryStream
  readonly trace: TraceLog

  constructor(
    initialMessages: HarnessMessage[],
    processInput: ProcessInput,
    queryStream: QueryStream,
    trace = new TraceLog(),
  ) {
    this.#messages = initialMessages
    this.#processInput = processInput
    this.#queryStream = queryStream
    this.trace = trace
  }

  getMessages(): readonly HarnessMessage[] {
    return [...this.#messages]
  }

  async *submitMessage(prompt: string): AsyncGenerator<HarnessMessage, void> {
    this.trace.record({
      type: 'call.entered',
      source: 'TraceableEngine.submitMessage',
      target: 'processInput',
    })
    const processed = await this.#processInput(prompt, [...this.#messages])
    this.#messages.push(...processed.messages)
    this.trace.record({
      type: 'state.mutated',
      owner: 'TraceableEngine',
      field: 'messages',
      size: this.#messages.length,
    })

    const requestView = [...this.#messages]
    this.trace.record({
      type: 'view.snapshotted',
      owner: 'TraceableEngine',
      field: 'requestView',
      size: requestView.length,
    })

    if (!processed.shouldQuery) {
      this.trace.record({
        type: 'branch.skipped',
        branch: 'queryStream',
        reason: 'processInput.shouldQuery=false',
      })
      return
    }

    this.trace.record({
      type: 'call.entered',
      source: 'TraceableEngine.submitMessage',
      target: 'queryStream',
    })
    try {
      for await (const message of this.#queryStream(requestView)) {
        this.trace.record({
          type: 'event.yielded',
          source: 'queryStream',
          eventKind: message.kind,
        })
        this.#messages.push(message)
        this.trace.record({
          type: 'state.mutated',
          owner: 'TraceableEngine',
          field: 'messages',
          size: this.#messages.length,
        })
        yield message
      }
    } catch (error) {
      this.trace.record({
        type: 'call.failed',
        source: 'queryStream',
        message: error instanceof Error ? error.message : String(error),
      })
      throw error
    }
  }
}

export type ToolProbe = {
  name: string
  run(): void
}

export async function passToolToPermission(
  tool: ToolProbe,
  canUseTool: (candidate: ToolProbe) => Promise<boolean>,
  trace: TraceLog,
): Promise<boolean> {
  trace.record({
    type: 'call.entered',
    source: 'permissionWrapper',
    target: 'canUseTool',
  })
  return canUseTool(tool)
}

export async function collect(
  stream: AsyncIterable<HarnessMessage>,
): Promise<HarnessMessage[]> {
  const messages: HarnessMessage[] = []
  for await (const message of stream) messages.push(message)
  return messages
}

File: curriculum/units/M04/code/python/trace_harness.py
from __future__ import annotations

from collections.abc import AsyncIterator, Awaitable, Callable, Sequence
from dataclasses import asdict, dataclass
from typing import Literal, TypedDict


class HarnessMessage(TypedDict):
    kind: Literal["user", "assistant"]
    text: str


@dataclass(frozen=True)
class TraceEvent:
    seq: int
    type: str
    source: str = ""
    target: str = ""
    owner: str = ""
    field: str = ""
    size: int = 0
    detail: str = ""


class TraceLog:
    def __init__(self) -> None:
        self.events: list[TraceEvent] = []

    def record(self, event_type: str, **details: str | int) -> None:
        self.events.append(
            TraceEvent(seq=len(self.events) + 1, type=event_type, **details)
        )

    def observed_call(self, source: str, target: str) -> bool:
        return any(
            event.type == "call.entered"
            and event.source == source
            and event.target == target
            for event in self.events
        )

    def to_dicts(self) -> list[dict[str, object]]:
        return [asdict(event) for event in self.events]


@dataclass(frozen=True)
class ProcessInputResult:
    messages: list[HarnessMessage]
    should_query: bool


ProcessInput = Callable[
    [str, Sequence[HarnessMessage]], Awaitable[ProcessInputResult]
]
QueryStream = Callable[[Sequence[HarnessMessage]], AsyncIterator[HarnessMessage]]


class TraceableEngine:
    def __init__(
        self,
        initial_messages: list[HarnessMessage],
        process_input: ProcessInput,
        query_stream: QueryStream,
        trace: TraceLog | None = None,
    ) -> None:
        self._messages = initial_messages
        self._process_input = process_input
        self._query_stream = query_stream
        self.trace = trace or TraceLog()

    def get_messages(self) -> list[HarnessMessage]:
        return list(self._messages)

    async def submit_message(self, prompt: str) -> AsyncIterator[HarnessMessage]:
        self.trace.record(
            "call.entered",
            source="TraceableEngine.submitMessage",
            target="processInput",
        )
        processed = await self._process_input(prompt, list(self._messages))
        self._messages.extend(processed.messages)
        self.trace.record(
            "state.mutated",
            owner="TraceableEngine",
            field="messages",
            size=len(self._messages),
        )

        request_view = list(self._messages)
        self.trace.record(
            "view.snapshotted",
            owner="TraceableEngine",
            field="requestView",
            size=len(request_view),
        )
        if not processed.should_query:
            self.trace.record(
                "branch.skipped",
                field="queryStream",
                detail="processInput.shouldQuery=false",
            )
            return

        self.trace.record(
            "call.entered",
            source="TraceableEngine.submitMessage",
            target="queryStream",
        )
        try:
            async for message in self._query_stream(request_view):
                self.trace.record(
                    "event.yielded",
                    source="queryStream",
                    detail=message["kind"],
                )
                self._messages.append(message)
                self.trace.record(
                    "state.mutated",
                    owner="TraceableEngine",
                    field="messages",
                    size=len(self._messages),
                )
                yield message
        except Exception as error:
            self.trace.record(
                "call.failed", source="queryStream", detail=str(error)
            )
            raise


@dataclass
class ToolProbe:
    name: str
    run: Callable[[], None]


async def pass_tool_to_permission(
    tool: ToolProbe,
    can_use_tool: Callable[[ToolProbe], Awaitable[bool]],
    trace: TraceLog,
) -> bool:
    trace.record(
        "call.entered", source="permissionWrapper", target="canUseTool"
    )
    return await can_use_tool(tool)


async def collect(stream: AsyncIterator[HarnessMessage]) -> list[HarnessMessage]:
    return [message async for message in stream]
